# Results and Discussion

## Scalability, Portability, and Efficiency

The re-implementation of EAGER into Nextflow offers a range of benefits over the
original custom pipeline framework.

Firstly, the new framework provides immediate integration of nf-core/eager into
various job schedulers in POSIX High-Performance-Cluster (HPC) environments,
cloud computing resources, as well as local workstations. This portability
allows users to run nf-core/eager regardless of the type of computing
infrastructure or cluster size (if applicable), with minimal effort or
configuration. This facilitates reproducibility and therefore maintenance of
standards within the field. This is further assisted by the in-built
compatibility with software environments and containers such as Conda
[@url:https://conda.io], Docker [@url:https://www.docker.com] and Singularity
[@url:https://sylabs.io]. These are isolated sandboxes that include all software
(with exact versions) required by the pipeline, in a form that is installable
and runnable by users regardless of the set up of their local software
environment. Another major change with nf-core/eager is that the
graphical-user-interface (GUI) set up of an EAGER run is now replaced with a
command-line-interface (CLI) as the primary user interaction mode. This is more
compatible and portable with most HPCs (that may not offer display of a window
system), and is in line with the vast majority of bioinformatic tools. We
therefore believe this will not be a hindrance to new researchers from outside
computational biology. However, a GUI-based pipeline set up is now offered via
the nf-core website ([https://nf-co.re/launch](https://nf-co.re/launch)), which
provides a commmon GUI format across multiple pipelines as well as additional
robusticity checks of input parameters for those less familiar with CLIs.
Typically the output of the launch functionality is a JSON file that can be used
with nf-core launch command as a single parameter (as with EAGER v1), however
upcoming integration with tower.nf [@url:https://tower.nf] will allow direct
submission of pipelines without any command line usage.

Secondly, reproducibility is made easier through the use of 'profiles' that can
define configuration parameters. These profiles can be managed at different
hierarchical levels. HPC-level profiles can specify parameters for the computing
environment (job schedulers, cache locations for containers, maximum memory and
CPU resources etc.), which can be centrally managed to ensure all users of a
group use the same settings. Pipeline-level profiles, specifying parameters for
nf-core/eager itself, allow fast access to routine pipeline-run parameters via a
single flag in the nf-core/eager run command, without having to configure each
new run from scratch. Compared to the original EAGER, which utilised per-FASTQ
XML files with hardcoded filepaths for a specific user's server, nf-core/eager
allows researchers to publish the specific profile used in their runs alongside
their publications, to ensure other groups can generate the same results. Usage
of profiles can also reduce mistakes caused by insufficient 'prose' based
reporting of program settings that can be regularly found in the literature. The
default nf-core/eager profile uses parameters evaluated in different
aDNA-specific contexts (e.g. in [@doi:10.1186/1471-2164-13-178]), and will be
updated in each new release as new studies are published.

Finally, nf-core/eager provides improved efficiency over the original EAGER
pipeline by replacing the sample-by-sample sequential processing with Nextflow's
asynchronous job parallelisation, whereby multiple pipeline steps and samples
are run in parallel (in addition to natively parallelised pipeline steps). This
is similar to the approach taken by paleomix, however nf-core/eager expands this
by utilising Nextflow's ability to customise resource settings the parameters for
every job in the pipeline; reducing unnecessary resource allocation that can
occur with unfamiliar users to each step of an NGS data processing pipeline.
This is particularly pertinent given the increasing use of centralised HPCs or
cloud computing that often use per-hour cost calculations.

## Updated Workflow

nf-core/eager follows a similar structural foundation to the original version of
EAGER and partially with paleomix. Given Illumina short-read FASTQ and/or BAM
files and a reference FASTA file, this can be split into four main stages:

1. Pre-processing
   - Sequencing quality control: FastQC
     [@url:https://www.bioinformatics.babraham.ac.uk/projects/fastqc]
   - Sequencing artefact clean-up (merging, adapter clipping): AdapterRemoval2
     [@doi:10.1186/s13104-016-1900-2]
   - Pre-processing statistics generation
2. Mapping and post-processing
   - Alignment against reference genome: BWA
     [@doi:10.1093/bioinformatics/btp324; @doi:10.1093/bioinformatics/btp698,
     @arxiv:1303.3997], CircularMapper [@doi:10.1186/s13059-016-0918-z], Bowtie2
     [@doi:10.1038/nmeth.1923]
   - Mapping quality filtering: SAMtools [@doi:10.1093/bioinformatics/btp352]
   - PCR duplicate removal: DeDup [@doi:10.1186/s13059-016-0918-z], Picard
     MarkDuplicates [@doi:10.1101/gr.107524.110]
   - Mapping statistics generation: PreSeq [@doi:10.1038/nmeth.2375], Qualimap2
     [@doi:10.1093/bioinformatics/btv566]
3. aDNA Evaluation and Modification
   - Damage profiling: DamageProfiler [@doi:10.5281/zenodo.3557708]
   - aDNA reads selection: PMDtools [@doi:10.1073/pnas.1318934111]
   - Damage removal: Bamutils[@doi:10.1101/gr.176552.114]
4. Genotyping and Consensus Sequence Generation: GATK
   [@doi:10.1101/gr.107524.110], ANGSD [@doi:10.1186/s12859-014-0356-4]
   VCF2Genome [@doi:10.1186/s13059-016-0918-z]

In nf-core/eager, all tools originally used in EAGER have been updated to their
latest versions, as available on Bioconda [@doi:10.1038/s41592-018-0046-7] and
Conda-forge [@url:https://conda-forge.github.io] to ensure widespread
accessibility and stability of utilised tools. The MapDamage2 (for damage
profile generation) [@doi:10.1093/bioinformatics/btt193] and Schmutzi (for
mitochondrial contamination estimation) [@doi:10.1186/s13059-015-0776-0] methods
have not been carried over to nf-core/eager, the first because a more performant
successor method is now available (DamageProfiler), and the latter because a
stable release of the method could not be migrated to Bioconda. We anticipate
that there will be an updated version of Schmutzi in the near future that will
allow us to integrate the method again in nf-core/eager, once a version is
released on Bioconda. As an alternative, estimation of human _nuclear_
contamination estimation is now offered through ANGSD
[@doi:10.1186/s12859-014-0356-4]. Support for the Bowtie2 aligner
[@doi:10.1038/nmeth.1923] has been updated to have default settings optimised
for aDNA [@doi:10.3389/fevo.2020.00105].

New tools to the basic workflow include fastp
[@doi:10.1093/bioinformatics/bty560] for the removal of 'poly-G' sequencing
artefacts that are common in 2-colour Illumina sequencing machines (such as the
increasingly popular NextSeq and NovaSeq platforms
[@url:https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/]).
For genotyping, we have now included FreeBayes [@arxiv:1207.3907] as an
alternative to the human-focused GATK tools, and have also added pileupCaller
[@url:https://github.com/stschiff/sequenceTools] for generation of genotyping
formats commonly utilised in ancient human population analysis. We have also
maintained the possibility of using the now unsupported GATK UnifiedGenotyper,
as the GATK HaplotypeCaller performs _de novo_ assembly around possible
variants, which may not be suitable for low-coverage aDNA data.

![Simplified schematic of the nf-core/eager workflow pipeline. Green filled
bubbles indicate new functionality added over the original EAGER
pipeline.](images/eager2_workflow.png){#fig:workflow-schematic
width="100%"}

We have further extended the functionality of the pipeline by adding ancient
metagenomic analysis, to identify the wider taxonomic content of a sample. We
have added the possibility to screen all off-target reads (not mapped to the
reference genome) with two metagenomic profilers: MALT [@doi:10.1101/050559;
@doi:10.1038/s41559-017-0446-6] and Kraken2 [@doi:10.1186/s13059-019-1891-0].
Characterisation of properties of authentic aDNA from MALT alignments is carried
out with the HOPS pipeline [@doi:10.1186/s13059-019-1903-0]. Ancient metagenomic
studies sometimes include comparative samples from living individuals
[@doi:10.1186/s40168-019-0717-3)]. To support open data, whilst respecting
personal data privacy, nf-core/eager includes a 'strip_fastq' script which
creates raw FASTQ files, but with all reads successfully mapped to the reference
genome removed. This allows safe upload of sequencing data to public
repositories after removal of identifiable (human) data.

| Functionality                              | EAGER | paleomix | nf-core/eager |
|--------------------------------------------|:-----:|:--------:|:-------------:|
| Command line set up                        | ✗     | ✓        | ✓             |
| GUI set up                                 | ✓     | ✗        | ✓             |
| Reproducible software environments offered | ✓     | ✗        | ✓             |
| HPC scheduler integration                  | ✗     | ✗        | ✓             |
| Cloud computing integration                | ✗     | ✗        | ✓             |
| Per-process resource optimisation          | ✗     | ~        | ✓             |
| Pipeline-step parallelisation              | ✗     | ✓        | ✓             |
| Sequencing lane merging                    | ✓     | ✓        | ✓             |
| Sequencing quality control                 | ✓     | ✗        | ✓             |
| Sequencing artefact removal                | ✗     | ✗        | ✓             |
| Adapter clipping/read merging              | ✓     | ✓        | ✓             |
| Post-processing sequencing QC              | ✗     | ✗        | ✓             |
| Reference mapping                          | ✓     | ✓        | ✓             |
| Reference mapping statistics               | ✓     | ✓        | ✓             |
| Multi-reference mapping                    | ✗     | ✓        | ✗             |
| Mapped reads filtering                     | ✓     | ✓        | ✓             |
| Off-target read treatment                  | ✓     | ✗        | ✓             |
| Off-target metagenomic profiling           | ✗     | ✗        | ✓             |
| Off-target metagenomic authentication      | ✗     | ✗        | ✓             |
| Library complexity estimation              | ✓     | ✗        | ✓             |
| Duplicate removal                          | ✓     | ✗        | ✓             |
| BAM merging                                | ✗     | ✓        | ✓             |
| Damage read filtering                      | ✓     | ✗        | ✓             |
| Contamination estimation (Human)           | ✓     | ✗        | ✓             |
| Biological sex determination (Human)       | ✗     | ✗        | ✓             |
| Genome coverage estimation                 | ✓     | ✓        | ✓             |
| Damage calculation                         | ✓     | ✓        | ✓             |
| Damage rescaling                           | ✗     | ✓        | ~             |
| SNP Calling/Genotyping                     | ✓     | ~        | ✓             |
| Consensus sequence generation              | ✓     | ~        | ✓             |
| Regions of interest statistics             | ~     | ✓        | ✓             |

Table: Comparison of usage and pipeline functionality of common ancient DNA
processing pipelines. Tick represents full functionality, tilde represents
partial functionality cross represents not implemented.
{#tbl:pipeline-comparison}.

Additional functionality tailored for ancient bacterial genomics includes
integration of a SNP alignment generation tool, MultiVCFAnalyzer
[@doi:10.1038/nature13591], which allows assessment of cross-mapping levels from
different related taxa to a reference genome - a common challenge in ancient
bacterial genome reconstruction [@doi:10.1146/annurev-genom-091416-035526]. The
output SNP alignment FASTA file can then be used for downstream analyses such as
phylogenetic tree construction. Simple coverage statistics of particular
annotations (e.g. genes) of an input reference is offered by bedtools
[@doi:10.1093/bioinformatics/btq033], which can be used in cases such as for
determining functional differences between ancient bacterial strains (as in
[@doi:10.1016/j.cub.2017.10.025]). When using a human reference genome,
nf-core/eager can now also can give estimates of the biological sex of a given
individual with Sex.DetERRmine [@doi:10.1038/s41467-018-07483-5]. A dedicated
'endogenous DNA' calculator (endorS.py) is also included, to provide a
percentage estimate of the sequenced reads matching the reference from the total
number of reads sequenced per library.

Given the large amount of sequencing often required to yield sufficient genome
coverage from aDNA data, palaeogeneticists tend to use multiple (differently
treated) libraries, and/or merge data from multiple sequencing runs per library.
The original EAGER pipeline could only run single libraries at a time, and in
these contexts required significant manual user input in merging different FASTQ
or BAM files. A major upgrade in nf-core/eager is that the new pipeline supports
automated processing of complex sequencing strategies for many samples. This is
facilitated by the optional use as input of a simple table in TSV format, which
includes file paths and additional metadata such as sample name, library name,
sequencing lane, colour chemistry, and UDG treatment. This allows simultaneous
processing and appropriate merging of heterogeneous data from multiple
sequencing runs and/or library types.

The original EAGER pipeline required users to look through many independent
output directories and files to make full assessment of their sequencing data.
This has now been replaced with a much more extensive MultiQC
[@doi:10.1093/bioinformatics/btw354] report. This tool aggregates the log files
of every supported tool into a single interactive report, and assists users in
making a fuller assessment of their sequencing and analysis runs. We have
strived to ensure that there is a corresponding MultiQC module for every tool
used by nf-core/eager where possible, to enable comprehensive evaluation of all
stages of the pipeline.

An overview of the entire pipeline is shown in Fig. {@fig:workflow-schematic},
and a tabular comparison of functionality between EAGER, paleomix and
nf-core/eager in {@tbl:pipeline-comparison}.

To demonstrate the simultaneous genomic analysis of human DNA and  metagenomic
screening for putative pathogen, and improved results reporting, we re-analysed
data from Barquera et al. [@doi:10.1016/j.cub.2020.04.002], who performed a
multi-discipline study of three 16th century individuals excavated from a mass
burial site in Mexico City. The authors reported genetic results showing
sufficient on-target human DNA (>1%) with typical aDNA damage (>20% C to T
reference mismatches in the first base of the 5' ends of reads) for downstream
population-genetic analysis, Y-chromosome coverage indicative that the three
individuals were genetically Male, and one individual (Lab ID: SJN003) contained
DNA suggesting a possible infection by _Treponema pallidum_, a species with a
variety of strains that can cause diseases such as syphilis, bejel and yaws, and
a second individual (Lab ID: SJN001) with the presence of Hepatitis B virus.
Both results were confirmed via in-solution enrichment approaches.

We were able to successfully replicate the human and pathogen screening results
in a single run of nf-core/eager. Mapping to the human reference genome (hs37d5)
with BWA aln and binning of off-target reads with MALT to the NCBI Nucleotide
database (2017-10-26) yielded the same results of all individuals having a
biological sex of Male as well as the same frequency of C -> T miscoding lesions
and short fragment lengths (characteristic of aDNA). Metagenomic hits to both
pathogens from the same individuals that yielded complete genomes in the
original publication were also detected. Both results and other processing
statistics were identified via a single interactive MultiQC report, excerpts of
which can be seen in Figure {@fig:dual-human-pathogen}. The full interactive
report can be seen in the supplementary information.

![Sections of MultiQC report (v1.10dev) from the outcome of simultaneous human
DNA and microbial pathogen screening with nf-core/eager, including **A**
Sex.DetERRmine output from Male biological sex assignment with coverages on X
and Y being half of that of autosomes, and **B** HOPS output with detection on
of both _Treponema pallidum_ and Hepatitis B virus reads. Other taxa in HOPS
output represent typical oral commensal microbiota found in teeth. MultiQC v1.10
will be included in a point release of nf-core/eager, once released by the
developer, but in the meantime can be manually run to get the same results as
v1.9 but with the integrate HOPS heatmap. Data was shotgun data from Barquera et
al. 2020 [@doi:10.1016/j.cub.2020.04.002], and replicated results here were
originally verified in the publication via enrichment methods. The full
interactive reports for both MultiQC 1.9 and 1.10 can be seen in the
supplementary
information.](images/dualhumanpathogen_exportedmqc_base_edit.png){#fig:dual-human-pathogen
width="70%"}

## Accessibility

Alongside the interactive MultiQC report, we have written extensive
documentation on all parts of running and interpreting the output of the
pipeline. Given that a large fraction of aDNA researchers come from fields
outside computational biology, and thus may have limited computational training,
we have written documentation [@url:https://nf-co.re/eager/docs] that also gives
guidance on how to interpret each section of the report - specifically in the
context of NGS and aDNA. This includes best practices or expected output
schematic images, which are published under CC-BY licenses to allow for use in
other training material (an example can be seen in Fig.
{@fig:doc-image-example}). We hope this open-access resource will make the study
of aDNA more accessible to researchers new to the field, by providing practical
guidelines on how to evaluate characteristics and effects of aDNA on downstream
analyses.

![Example schematic images of pipeline output documentation that can assist new
users in the interpretation to next-generation-sequencing aDNA
processing.](images/fastqc_adapter_content.png){#fig:doc-image-example
width="70%"}

The development of nf-core/eager in Nextflow and the nf-core initiative will
also improve open-source development, while ensuring the high quality of
community contributions to the pipeline. While Nextflow is written primarily in
Groovy, the Nextflow DSL simplifies a number of concepts to an intermediate
level that bioinformaticians without Java/Groovy experience can easily access
(regardless of own programming language experience). Furthermore, Nextflow
places ubiquitous and more widely known command-line interfaces, such as bash,
in a prominent position within the code, rather than custom java code and
classes. We hope this will motivate further bug fixes and feature contributions
from the community, to keep the pipeline state-of-the-art and ensure a longer
life-cycle. This will also be supported by the active and welcoming nf-core
community who provide general guidance and advice on developing Nextflow and
nf-core pipelines.

## Comparisons with other pipelines

We compared pipeline run-times of the three pipelines to show that the new
implementation of nf-core/eager is equivalent or more efficient than EAGER or
paleomix. While similar pipelines designed for aDNA have also been released,
such as ATLAS [@doi:10.1101/105346] or HOPS [@doi:10.1186/s13059-019-1903-0],
these generally have been designed with specific contexts in mind (e.g. human
population genetics or pathogen screening) and do not include many of the
preprocessing steps required for analysing NGS data.

We ran each pipeline on a subset of Viking-age cod (_Gadus morhua_) genomic data
from [@doi:10.1073/pnas.1710186114]. This data was originally run using
paleomix, and was re-run here as described but with paleomix (v1.2.14), with
settings for the other two pipelines as close as possible to the original paper
(EAGER with v1.92.33, and nf-core/EAGER with v2.2.0dev, commit
830c22d448441e5e19508c198f530a7656c9f25d). Benchmarking environment and exact
pipeline run settings can be seen in Methods and Supplementary methods. Two samples from
three Illumina paired-end sequencing runs were analysed, with adapter clipping
and merging (AdapterRemoval), mapping (bwa aln), duplicate removal
(markduplicates) and damage profiling (paleomix: mapDamage2, EAGER and
nf-core/EAGER: DamageProfiler) steps performed. We ran the commands for each
tool sequentially, but repeated these batch of commands 10 times - to account
for variability in the cloud service's IO connection. Run times were measured
using the GNU time tool (v1.7).

|Pipeline                  |Version  |Environment  |real        |sys        |user          |
|:-------------------------|:--------|:------------|:-----------|:----------|:-------------|
|nf-core-eager (optimised) |2.2.0dev |singularity  |105.6 ± 4.6 |13.6 ± 0.7 |1593 ± 79.7   |
|paleomix (optimised)      |1.2.14   |conda        |130.6 ± 8.7 |12 ± 0.7   |1820.2 ± 36.9 |
|nf-core-eager             |2.2.0dev |singularity  |209.2 ± 4.4 |11 ± 0.9   |1407.7 ± 30.2 |
|EAGER                     |1.92.37  |singularity  |224.2 ± 4.9 |22.9 ± 0.3 |1736.3 ± 70.2 |
|paleomix                  |1.2.14   |conda        |314.6 ± 2.9 |10.7 ± 1   |1506.7 ± 14   |

Table: Comparison of run times in minutes between three ancient DNA pipelines.
paleomix and nf-core/eager has additional runs with 'optimised' parameters with
fairer computational resources matching modern multi-threading strategies for
comparison. Values represent mean and standard deviation of run times in minutes
{#tbl:benchmarking-runtimes}.

A summary of runtimes of the benchmarking tests can be seen in Table
{@tbl:benchmarking-runtimes}. nf-core/eager showed lowest runtimes across all
three time metrics (Real: real time, System: cumulative CPU system-task times,
User: cumulative CPU time of all tasks) when running on default parameters. This
highlights the improved efficiency of nf-core/eager's asynchronous processing
system and per-process resource customisation (here represented by nf-core/eager
defaults designed for typical HPC set ups).

As a more realistic demonstration of modern computing multi-threading set ups,
thread, we also re-ran paleomix with the flag --max-bwa-threads set to 4 (listed
in Table {@tbl:benchmarking-runtimes} as 'optimised'), which is equivalent to a
single bwa process of nf-core/eager. This resulted in a much faster run-time
than that of default nf-core/eager, due to the approach of paleomix of mapping
each lane of a library separately, whereas nf-core/eager will map all lanes of a
single library merged together. Therefore, given that each library was split
across three lanes, increasing the threads of bwa to 4 resulted in 12 per
library, whereas nf-core/eager only gave 4 for a single bwa process of one
library. While the paleomix approach is valid, we opted to retain the
per-library mapping as it is often the longest running step of NGS
genome-mapping pipelines, and it prevents flooding of HPC scheduling systems
with many long-running jobs. Secondly, if users regularly use multi-lane data,
due nf-core/eager's fine-granularity control, they can simply modify
nf-core/eager's bwa process resources via config files to account for this. When
we optimised parameters were used for bwa multi-threading and multiple lanes to
the same number of bwa threads as the optimised paleomix run, nf-core/eager
still displayed shorter runtimes.  All metrics including mapped reads,
percentage on-target, mean depth coverage and mean read lengths across all
pipelines were extremely similar across all pipelines and replicates (see
methods {@tbl:benchmarking-results}).
