---
author-meta:
- James A. Fellows Yates
- Thiseas C. Lamnidis
- Maxime Borry
- "Aida Andrades Valtue\xF1a"
- "Zandra Fagern\xE4s"
- Stephen Clayton
- Maxime U. Garcia
- Judith Neukamm
- Alexander Peltzer
bibliography:
- content/manual-references.json
date-meta: '2020-10-13'
header-includes: "<!--\nManubot generated metadata rendered from header-includes-template.html.\nSuggest improvements at https://github.com/manubot/manubot/blob/master/manubot/process/header-includes-template.html\n-->\n<meta name=\"dc.format\" content=\"text/html\" />\n<meta name=\"dc.title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta name=\"citation_title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta property=\"og:title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta property=\"twitter:title\" content=\"Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager\" />\n<meta name=\"dc.date\" content=\"2020-10-13\" />\n<meta name=\"citation_publication_date\" content=\"2020-10-13\" />\n<meta name=\"dc.language\" content=\"en-GB\" />\n<meta name=\"citation_language\" content=\"en-GB\" />\n<meta name=\"dc.relation.ispartof\" content=\"Manubot\" />\n<meta name=\"dc.publisher\" content=\"Manubot\" />\n<meta name=\"citation_journal_title\" content=\"Manubot\" />\n<meta name=\"citation_technical_report_institution\" content=\"Manubot\" />\n<meta name=\"citation_author\" content=\"James A. Fellows Yates\" />\n<meta name=\"citation_author_institution\" content=\"Microbiome Sciences Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_institution\" content=\"Institut f\xFCr Vor- und Fr\xFChgeschichtliche Arch\xE4ologie und Provinzialr\xF6mische Arch\xE4ologie, Ludwig Maximilian University, Munich, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-5585-6277\" />\n<meta name=\"twitter:creator\" content=\"@jafellowsyates\" />\n<meta name=\"citation_author\" content=\"Thiseas C. Lamnidis\" />\n<meta name=\"citation_author_institution\" content=\"Population Genetics Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-4485-8570\" />\n<meta name=\"twitter:creator\" content=\"@TCLamnidis\" />\n<meta name=\"citation_author\" content=\"Maxime Borry\" />\n<meta name=\"citation_author_institution\" content=\"Microbiome Sciences Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-9140-7559\" />\n<meta name=\"twitter:creator\" content=\"@notmaxib\" />\n<meta name=\"citation_author\" content=\"Aida Andrades Valtue\xF1a\" />\n<meta name=\"citation_author_institution\" content=\"Computational Pathogenomics Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-1737-2228\" />\n<meta name=\"twitter:creator\" content=\"@aidaanva\" />\n<meta name=\"citation_author\" content=\"Zandra Fagern\xE4s\" />\n<meta name=\"citation_author_institution\" content=\"Microbiome Sciences Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-2667-3556\" />\n<meta name=\"twitter:creator\" content=\"@ZandraSelina\" />\n<meta name=\"citation_author\" content=\"Stephen Clayton\" />\n<meta name=\"citation_author_institution\" content=\"Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-5223-9695\" />\n<meta name=\"citation_author\" content=\"Maxime U. Garcia\" />\n<meta name=\"citation_author_institution\" content=\"Department of Oncology-Pathology, Karolinska Institutet, Stockholm, Sweden\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0003-2827-9261\" />\n<meta name=\"twitter:creator\" content=\"@gau\" />\n<meta name=\"citation_author\" content=\"Judith Neukamm\" />\n<meta name=\"citation_author_institution\" content=\"Palaeogenetics Group, Institute of Evolutionary Medicine, University of Zurich, Z\xFCrich, Switzerland\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0001-8141-566X\" />\n<meta name=\"twitter:creator\" content=\"@JudithNeukamm\" />\n<meta name=\"citation_author\" content=\"Alexander Peltzer\" />\n<meta name=\"citation_author_institution\" content=\"Quantitative Biology Center (QBiC), Eberhard-Karls-Universit\xE4t, T\xFCbingen, Germany\" />\n<meta name=\"citation_author_orcid\" content=\"0000-0002-6503-2180\" />\n<meta name=\"twitter:creator\" content=\"@alex_peltzer\" />\n<link rel=\"canonical\" href=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta property=\"og:url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta property=\"twitter:url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta name=\"citation_fulltext_html_url\" content=\"https://apeltzer.github.io/eager2-paper/\" />\n<meta name=\"citation_pdf_url\" content=\"https://apeltzer.github.io/eager2-paper/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"application/pdf\" href=\"https://apeltzer.github.io/eager2-paper/manuscript.pdf\" />\n<link rel=\"alternate\" type=\"text/html\" href=\"https://apeltzer.github.io/eager2-paper/v/201087cc28e29d7f6e96e327f29fea6a072581b8/\" />\n<meta name=\"manubot_html_url_versioned\" content=\"https://apeltzer.github.io/eager2-paper/v/201087cc28e29d7f6e96e327f29fea6a072581b8/\" />\n<meta name=\"manubot_pdf_url_versioned\" content=\"https://apeltzer.github.io/eager2-paper/v/201087cc28e29d7f6e96e327f29fea6a072581b8/manuscript.pdf\" />\n<meta property=\"og:type\" content=\"article\" />\n<meta property=\"twitter:card\" content=\"summary_large_image\" />\n<meta property=\"og:image\" content=\"https://github.com/apeltzer/eager2-paper/raw/201087cc28e29d7f6e96e327f29fea6a072581b8/content/images/nf-core-eager_social_preview.png\" />\n<meta property=\"twitter:image\" content=\"https://github.com/apeltzer/eager2-paper/raw/201087cc28e29d7f6e96e327f29fea6a072581b8/content/images/nf-core-eager_social_preview.png\" />\n<link rel=\"icon\" type=\"image/png\" sizes=\"192x192\" href=\"https://manubot.org/favicon-192x192.png\" />\n<link rel=\"mask-icon\" href=\"https://manubot.org/safari-pinned-tab.svg\" color=\"#ad1457\" />\n<meta name=\"theme-color\" content=\"#ad1457\" />\n<!-- end Manubot generated metadata -->"
keywords:
- nf-core
- nextflow
- ancientDNA
- aDNA
- palaeogenetics
- archaeogenetics
- bioinformatics
lang: en-GB
manubot-clear-requests-cache: false
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
title: Reproducible, portable, and efficient ancient genome reconstruction with nf-core/eager
...






<small><em>
This manuscript
([permalink](https://apeltzer.github.io/eager2-paper/v/201087cc28e29d7f6e96e327f29fea6a072581b8/))
was automatically generated
from [apeltzer/eager2-paper@201087c](https://github.com/apeltzer/eager2-paper/tree/201087cc28e29d7f6e96e327f29fea6a072581b8)
on October 13, 2020.
</em></small>

## Authors



+ **James A. Fellows Yates**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5585-6277](https://orcid.org/0000-0001-5585-6277)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [jfy133](https://github.com/jfy133)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [jafellowsyates](https://twitter.com/jafellowsyates)<br>
  <small>
     Microbiome Sciences Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany; Institut für Vor- und Frühgeschichtliche Archäologie und Provinzialrömische Archäologie, Ludwig Maximilian University, Munich, Germany
     · Funded by Max Planck Society
  </small>

+ **Thiseas C. Lamnidis**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-4485-8570](https://orcid.org/0000-0003-4485-8570)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [TCLamnidis](https://github.com/TCLamnidis)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [TCLamnidis](https://twitter.com/TCLamnidis)<br>
  <small>
     Population Genetics Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Maxime Borry**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-9140-7559](https://orcid.org/0000-0001-9140-7559)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [maxibor](https://github.com/maxibor)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [notmaxib](https://twitter.com/notmaxib)<br>
  <small>
     Microbiome Sciences Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Aida Andrades Valtueña**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-1737-2228](https://orcid.org/0000-0002-1737-2228)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [aidaanva](https://github.com/aidaanva)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [aidaanva](https://twitter.com/aidaanva)<br>
  <small>
     Computational Pathogenomics Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Zandra Fagernäs**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-2667-3556](https://orcid.org/0000-0003-2667-3556)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [ZandraFagernas](https://github.com/ZandraFagernas)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [ZandraSelina](https://twitter.com/ZandraSelina)<br>
  <small>
     Microbiome Sciences Group, Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Stephen Clayton**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5223-9695](https://orcid.org/0000-0001-5223-9695)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [sc13-bioinf](https://github.com/sc13-bioinf)<br>
  <small>
     Department of Archaeogenetics, Max Planck Institute for the Science of Human History, Jena, Germany
     · Funded by Max Planck Society
  </small>

+ **Maxime U. Garcia**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-2827-9261](https://orcid.org/0000-0003-2827-9261)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [MaxUlysse](https://github.com/MaxUlysse)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [gau](https://twitter.com/gau)<br>
  <small>
     Department of Oncology-Pathology, Karolinska Institutet, Stockholm, Sweden
     · Funded by Barncancerfonden
  </small>

+ **Judith Neukamm**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-8141-566X](https://orcid.org/0000-0001-8141-566X)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [JudithNeukamm](https://github.com/JudithNeukamm)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [JudithNeukamm](https://twitter.com/JudithNeukamm)<br>
  <small>
     Palaeogenetics Group, Institute of Evolutionary Medicine, University of Zurich, Zürich, Switzerland
  </small>

+ **Alexander Peltzer**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-6503-2180](https://orcid.org/0000-0002-6503-2180)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [apeltzer](https://github.com/apeltzer)
    · ![Twitter icon](images/twitter.svg){.inline_icon}
    [alex_peltzer](https://twitter.com/alex_peltzer)<br>
  <small>
     Quantitative Biology Center (QBiC), Eberhard-Karls-Universität, Tübingen, Germany
  </small>



## Abstract {.page_break_before}

The broadening utilisation of ancient DNA to address archaeological,
palaeontological, and biological questions is resulting in a rising diversity in
the size of laboratories and scale of analyses being performed. In the context
of this heterogeneous landscape, we present nf-core/eager, an advanced and
entirely redesigned and extended version of the EAGER pipeline for the analysis of ancient genomic
data. This Nextflow pipeline aims to address three main themes: accessibility
and adaptability to different computing configurations, reproducibility to
ensure robust analytical standards, and updating the pipeline to
the latest routine ancient genomic practises. This new version of EAGER has been
developed within the nf-core initiative to ensure high-quality software
development and maintenance support; contributing to a long-term lifecycle for
the pipeline. nf-core/eager will assist in ensuring that ancient DNA sequencing
data can be used by a diverse range of research groups and fields.


## Introduction

Ancient DNA (aDNA) has become a widely accepted source of biological data,
helping to provide new perspectives for a range of fields including archaeology,
cultural heritage, evolutionary biology, ecology, and palaeontology. The
utilisation of short-read high-throughput sequencing has allowed the recovery of
whole genomes and genome-wide data from a wide variety of sources, including
(but not limited to), the skeletal remains of animals
[@doi:10.1016/j.cub.2015.04.007; @doi:10.1038/nature12323;
@doi:10.1073/pnas.1901169116; @doi:10.1073/pnas.1710186114], modern and archaic
humans [@doi:10.1038/s41586-018-0094-2; @doi:10.1126/science.1188021;
@doi:10.1126/science.1224344; @doi:10.1038/s41586-018-0455-x], bacteria
[@doi:10.1038/nature13591; @doi:10.1073/pnas.1812865115;
@doi:10.1371/journal.ppat.1006997], viruses [@doi:10.1038/s41586-018-0097-z;
@doi:10.7554/eLife.36666], plants [@doi:10.1111/eva.12594;
@doi:10.1038/s41559-019-0921-3], palaeofaeces [@doi:10.1016/j.chom.2019.08.018;
@doi:10.7717/peerj.9001], dental calculus [@doi:10.1038/ng.2906;
@doi:10.1038/nature21674], sediments [@doi:10.1038/nature12921;
@doi:10.1126/science.aam9695], medical slides [@doi:10.1093/molbev/msz264],
parchment [@doi:10.1098/rstb.2013.0379], and recently, ancient ‘chewing gum’
[@doi:10.1038/s41467-019-13549-9; @doi:10.1038/s42003-019-0399-1]. Improvement
in laboratory protocols to increase yields of otherwise trace amounts of DNA has
at the same time led to studies that can total hundreds of ancient individuals
[@doi:10.1038/nature25738; @doi:10.1038/nature25778], spanning single
[@doi:10.1038/nature10549] to thousands of organisms [@doi:10.1038/ng.2906].
These differences of disciplines have led to a heterogeneous landscape in terms
of the types of analyses undertaken, and their computational resource
requirements [@doi:10.1093/bib/bbu022; @doi:10.3389/fgene.2018.00575]. Taking
into consideration the unequal distribution of resources (and infrastructure
such as internet connection), easy-to-deploy, streamlined and efficient
pipelines can help increase accessibility to high-quality analyses.

The degraded nature of aDNA poses an extra layer of complexity to standard
modern genomic analysis. Through a variety of processes [@doi:10.1038/362709a0]
DNA molecules fragment over time, resulting in ultra-short molecules
[@doi:10.1038/nature17405]. These sequences have low nucleotide complexity
making it difficult to identify with precision which part of the genome a read
(a sequenced DNA molecule) is derived from. Fragmentation without a 'clean
break' leads to uneven ends, consisting of single-stranded 'overhangs' at end of
molecules, which are susceptible to chemical processes such as deamination of
nucleotides. These damaged nucleotides then lead to misincorporation of
complementary bases during library construction for high-throughput DNA
sequencing [@doi:10.1073/pnas.0704665104]. On top of this, taphonomic processes
such as heat, moisture, and microbial- and burial-environment processes lead to
varying rates of degradation [@doi:10.1093/nar/gkx361;
@doi:10.1146/annurev-genom-091416-035526]. The original DNA content of a sample
is therefore increasingly lost over time and supplanted by younger
'environmental' DNA. Later handling by archaeologists, museum curators, and
other researchers can also contribute 'modern' contamination. While these
characteristics can help provide evidence towards the 'authenticity' of true
aDNA sequences (e.g. the aDNA cytosine to thymine or C to T 'damage'
deamination profiles [@doi:10.1093/bioinformatics/btt193]), they also pose
specific challenges for genome reconstruction, such as unspecific DNA alignment
and/or low coverage and miscoding lesions that can result in low-confidence
genotyping. These factors often lead to prohibitive sequencing costs when
retrieving enough data for modern high-throughput short-read sequencing data
pipelines (such as more than 1 billion reads for a 1X depth coverage _Yersinia
pestis_ genome [@doi:10.1016/j.cell.2015.10.009]), and thus aDNA-tailored
methods and techniques are required to overcome these challenges.

Two previously published and commonly used pipelines in the field are PALEOMIX
[@doi:10.1038/nprot.2014.063] and EAGER [@doi:10.1186/s13059-016-0918-z]. These
two pipelines take a similar approach to link together standard tools used for
Illumina high-throughput short-read data processing (sequencing quality control,
sequencing adapter removal/and or paired-end read merging, mapping of reads to a
reference genome, genotyping, etc.). However, they have a specific focus on
tools that are designed for, or well-suited for aDNA (such as the bwa aln
algorithm for ultra-short molecules [@doi:10.1093/bioinformatics/btp324] and
mapDamage [@doi:10.1093/bioinformatics/btr347] for evaluation of aDNA
characteristics). Yet, neither of these genome reconstruction pipelines have had
major updates to bring them in-line with current routine aDNA analyses.
*Meta*genomic screening of off-target genomic reads for pathogens or microbiomes
[@doi:10.1038/ng.2906; @doi:10.1038/nature21674] has become particularly common
in palaeo- and archaeogenetics, given its role in revealing widespread
infectious disease and possible epidemics that had previously been undetected in
the archaeological record [@doi:10.1016/j.cell.2015.10.009;
@doi:10.1016/j.cub.2017.10.025; @doi:10.1038/s41586-018-0097-z;
@doi:10.7554/eLife.36666]. Without easy access to the latest field-established
analytical routines, ancient genome studies risk being published without the
necessary quality control checks that ensure aDNA authenticity as well as
limiting the full range of possibilities from their data. Given that material
from samples is limited, there are both ethical as well as economical interests
to maximise analytical yield [@doi:10.3390/genes8070180].

To address these shortcomings, we have completely re-implemented the latest
version of the EAGER pipeline in Nextflow [@doi:10.1038/nbt.3820] (a
domain-specific-language or 'DSL', specifically designed for the construction of
omics analysis pipelines), introduced new features, and more flexible pipeline
configuration. In addition, the renamed pipeline - nf-core/eager
\- has been developed in the context of the nf-core community framework
[@doi:10.1038/s41587-020-0439-x], which enforces strict guidelines for
best-practices in software development.


# Results and Discussion

## Scalability, Portability, and Efficiency

The re-implementation of EAGER into Nextflow offers a range of benefits over the
original custom pipeline framework.

Firstly, the new framework provides immediate integration of nf-core/eager into
various job schedulers in POSIX High-Performance-Cluster (HPC) environments,
cloud computing resources, as well as local workstations. This portability
allows users to set up nf-core/eager regardless of the type of computing
infrastructure or cluster size (if applicable), with minimal effort or
configuration. This facilitates reproducibility and therefore maintenance of
standards within the field. Portability is further assisted by the in-built
compatibility with software environments and containers such as Conda
[@url:https://conda.io], Docker [@url:https://www.docker.com] and Singularity
[@url:https://sylabs.io]. These are isolated software 'sandbox' environments
that include all software (with exact versions) required by the pipeline, in a
form that is installable and runnable by users regardless of the set up of their
local software environment. Another major change with nf-core/eager is that the
primary user interaction mode of a pipeline run set up is now with a
command-line interface (CLI), replacing the graphical-user-interface (GUI) of
the original EAGER pipeline. This is more portable and compatible with most HPCs (that
may not offer display of a window system), and is in line with the vast majority
of bioinformatics tools. We therefore believe this will not be a hindrance to
new researchers from outside computational biology. However, a GUI-based
pipeline set up is still avaliable via the nf-core website's Launch page
([https://nf-co.re/launch](https://nf-co.re/launch)), which provides a common
GUI format across multiple pipelines as well as additional robustness checks of
input parameters for those less familiar with CLIs. Typically the output of the
launch functionality is a JSON file that can be used with a nf-core/tools launch
command as a single parameter (similar to the original EAGER), however
integration with Nextflow's companion monitoring tool tower.nf
[@url:https://tower.nf] also allows direct submission of pipelines without any
command line usage.

Secondly, reproducibility is made easier through the use of 'profiles' that can
define configuration parameters. These profiles can be managed at different
hierarchical levels. *HPC-level profiles* can specify parameters for the computing
environment (job schedulers, cache locations for containers, maximum memory and
CPU resources etc.), which can be centrally managed to ensure all users of a
group use the same settings. *Pipeline-level profiles*, specifying parameters for
nf-core/eager itself, allow fast access to routine pipeline-run parameters via a
single flag in the nf-core/eager run command, without having to configure each
new run from scratch. Compared to the original EAGER, which utilised per-FASTQ
XML files with hardcoded filepaths for a specific user's server, nf-core/eager
allows researchers to publish the specific profile used in their runs alongside
their publications, that can also be used by other groups to generate the same
results. Usage of profiles can also reduce mistakes caused by insufficient
'prose' based reporting of program settings that can be regularly found in the
literature. The default nf-core/eager profile uses parameters evaluated in
different aDNA-specific contexts (e.g. in [@doi:10.1186/1471-2164-13-178]), and
will be updated in each new release as new studies are published.

Finally, nf-core/eager provides improved efficiency over the original EAGER
pipeline by replacing the sample-by-sample sequential processing with Nextflow's
asynchronous job parallelisation, whereby multiple pipeline steps and samples
are run in parallel (in addition to natively parallelised pipeline steps). This
is similar to the approach taken by PALEOMIX, however nf-core/eager expands this
by utilising Nextflow's ability to customise the resource parameters for every
job in the pipeline; reducing unnecessary resource allocation that can occur
with unfamiliar users to each step of a high-throughput short-read data
processing pipeline. This is particularly pertinent given the increasing use of
centralised HPCs or cloud computing that often use per-hour cost calculations.

## Updated Workflow

nf-core/eager follows a similar structural foundation to the original version of
EAGER and partially to PALEOMIX. Given Illumina short-read FASTQ and/or BAM
files and a reference FASTA file, the core functionality of nf-core/eager can be split in five
main stages:

1. Pre-processing
   - Sequencing quality control: FastQC
     [@url:https://www.bioinformatics.babraham.ac.uk/projects/fastqc]
   - Sequencing artefact clean-up (merging, adapter clipping): AdapterRemoval2
     [@doi:10.1186/s13104-016-1900-2], fastp [@doi:10.1093/bioinformatics/bty560
   - Pre-processing statistics generation: FastQC
2. Mapping and post-processing
   - Alignment against reference genome: BWA aln and mem
     [@doi:10.1093/bioinformatics/btp324; @doi:10.1093/bioinformatics/btp698,
     @arxiv:1303.3997], CircularMapper [@doi:10.1186/s13059-016-0918-z], Bowtie2
     [@doi:10.1038/nmeth.1923]
   - Mapping quality filtering: SAMtools [@doi:10.1093/bioinformatics/btp352]
   - PCR duplicate removal: DeDup [@doi:10.1186/s13059-016-0918-z], Picard
     MarkDuplicates [@doi:10.1101/gr.107524.110]
   - Mapping statistics generation: SAMTools, PreSeq [@doi:10.1038/nmeth.2375], Qualimap2
     [@doi:10.1093/bioinformatics/btv566], bedtools [@doi:10.1093/bioinformatics/btq033], Sex.DetERRmine [@doi:10.1038/s41467-018-07483-5]
3. aDNA evaluation and modification
   - Damage profiling: DamageProfiler [@doi:10.1101/2020.10.01.322206]
   - aDNA reads selection: PMDtools [@doi:10.1073/pnas.1318934111]
   - Damage removal/Base trimming: Bamutils[@doi:10.1101/gr.176552.114]
   - Human nuclear contamination estimation: ANGSD [@doi:10.1186/s12859-014-0356-4]
4. Variant calling and consensus sequence generation: GATK UnifiedGenotyper and HaploTypeCaller
   [@doi:10.1101/gr.107524.110], ANGSD [@doi:10.1186/s12859-014-0356-4], sequenceTools pileupCaller [@url:https://github.com/stschiff/sequenceTools]
   VCF2Genome [@doi:10.1186/s13059-016-0918-z], MultiVCFAnalyzer @doi:10.1038/nature13591]
5. Report generation: MultiQC [@doi:10.1093/bioinformatics/btw354]

In nf-core/eager, all tools originally used in EAGER have been updated to their
latest versions, as available on Bioconda [@doi:10.1038/s41592-018-0046-7] and
Conda-forge [@url:https://conda-forge.github.io], to ensure widespread
accessibility and stability of utilised tools. The MapDamage2 (for damage
profile generation) [@doi:10.1093/bioinformatics/btt193] and Schmutzi (for
mitochondrial contamination estimation) [@doi:10.1186/s13059-015-0776-0] methods
have not been carried over to nf-core/eager, the first because a more performant
successor method is now available (DamageProfiler), and the latter because a
stable release of the method could not be migrated to Bioconda. We anticipate
that there will be an updated version of Schmutzi in the near future that will
allow us to integrate the method again into nf-core/eager. As an alternative,
estimation of human _nuclear_ contamination is now offered through
ANGSD [@doi:10.1186/s12859-014-0356-4]. Support for the Bowtie2 aligner
[@doi:10.1038/nmeth.1923] has been updated to have default settings optimised
for aDNA [@doi:10.3389/fevo.2020.00105].

New tools to the basic workflow include fastp
[@doi:10.1093/bioinformatics/bty560] for the removal of 'poly-G' sequencing
artefacts that are common in 2-colour Illumina sequencing machines (such as the
increasingly popular NextSeq and NovaSeq platforms
[@url:https://sequencing.qcfail.com/articles/illumina-2-colour-chemistry-can-overcall-high-confidence-g-bases/]).
For variant calling, we have now included FreeBayes [@arxiv:1207.3907] as an
alternative to the human-focused GATK tools, and have also added pileupCaller
[@url:https://github.com/stschiff/sequenceTools] for generation of genotyping
formats commonly utilised in ancient human population analysis. We have also
maintained the possibility of using the now unsupported GATK UnifiedGenotyper,
as the supported replacement GATK HaplotypeCaller performs _de novo_ assembly
around possible variants; something that may not be suitable for low-coverage
aDNA data.

![Simplified schematic of the nf-core/eager workflow pipeline. Green filled
bubbles indicate new functionality added over the original EAGER
pipeline.](images/eager2_workflow.png){#fig:workflow-schematic width="90%"}

Additional functionality tailored for ancient bacterial genomics includes
integration of a SNP alignment generation tool, MultiVCFAnalyzer
[@doi:10.1038/nature13591], which allows assessment of cross-mapping levels from
different related taxa to a reference genome - a common challenge in ancient
bacterial genome reconstruction [@doi:10.1146/annurev-genom-091416-035526]. The
output SNP consensus alignment FASTA file can then be used for downstream
analyses such as phylogenetic tree construction. Simple coverage statistics of
particular annotations (e.g. genes) of an input reference is offered by bedtools
[@doi:10.1093/bioinformatics/btq033], which can be used in cases such as for
providing initial indications of functional differences between ancient
bacterial strains (as in [@doi:10.1016/j.cub.2017.10.025]). When using a human
reference genome, nf-core/eager can also give estimates of the relative coverage
on the X and Y chromosomes with Sex.DetERRmine that can be used to infer the
biological sex of a given human individual [@doi:10.1038/s41467-018-07483-5]. A
dedicated 'endogenous DNA' calculator (endorS.py) is also included, to provide a
percentage estimate of the sequenced reads matching the reference ('on-target')
from the total number of reads sequenced per library.

Given the large amount of sequencing often required to yield sufficient genome
coverage from aDNA data, palaeogeneticists tend to use multiple (differently
treated) libraries, and/or merge data from multiple sequencing runs of each
library or even samples. The original EAGER pipeline could only run a single
library at a time, and in these contexts required significant manual user input
in merging different FASTQ or BAM files of related libraries. A major upgrade in
nf-core/eager is that the new pipeline supports automated processing of complex
sequencing strategies for many samples, similar to PALEOMIX. This is facilitated
by the optional use of a simple table table (in TSV format, a format more
commonly used in wet-lab stages of data generation, compared to PALEOMIX's YAML
format) that includes file paths and additional metadata such as sample name,
library name, sequencing lane, colour chemistry, and UDG treatment. This allows
automated and simultaneous processing and appropriate merging and treatment of
heterogeneous data from multiple sequencing runs and/or library types.

The original EAGER  and PALEOMIX pipelines required users to look through many
independent output directories and files to make full assessment of their
sequencing data. This has now been replaced in nf-core/eager with a much more
extensive MultiQC report [@doi:10.1093/bioinformatics/btw354]. This tool
aggregates the log files of every supported tool into a single interactive
report, and assists users in making a fuller assessment of their sequencing and
analysis runs. We have developed a corresponding MultiQC module for every tool
used by nf-core/eager, where possible, to enable comprehensive evaluation of all
stages of the pipeline.

We have further extended the functionality of the original EAGER pipeline by
adding ancient metagenomic analysis; allowing reconstruction of the wider
taxonomic content of a sample. We have added the possibility to screen all
off-target reads (not mapped to the reference genome) with two metagenomic
profilers: MALT [@doi:10.1101/050559; @doi:10.1038/s41559-017-0446-6] and
Kraken2 [@doi:10.1186/s13059-019-1891-0], in parallel to the mapping to a given
reference genome (typically of the host individual, assuming the sample is from
a skeleton). Characterisation of properties of authentic aDNA from metagenomic
MALT alignments is carried out with MaltExtract of the HOPS pipeline
[@doi:10.1186/s13059-019-1903-0]. This functionality can be used either for
microbiome screening or putative pathogen detection. Ancient metagenomic studies
sometimes include comparative samples from living individuals
[@doi:10.1186/s40168-019-0717-3)]. To support open data, whilst respecting
personal data privacy, nf-core/eager includes a 'FASTQ host removal' script
which creates raw FASTQ files, but with all reads successfully mapped to the
reference genome removed. This allows safe upload of metagenomic non-host
sequencing data to public repositories after removal of identifiable (human)
data, for example for microbiome studies.

An overview of the entire pipeline is shown in Figure {@fig:workflow-schematic},
and a tabular comparison of functionality between EAGER, PALEOMIX and
nf-core/eager in Table {@tbl:pipeline-comparison}.

To demonstrate the simultaneous genomic analysis of human DNA and metagenomic
screening for putative pathogens, as well as improved results reporting, we
re-analysed data from Barquera et al. 2020 [@doi:10.1016/j.cub.2020.04.002], who
performed a multi-discipline study of three 16th century individuals excavated
from a mass burial site in Mexico City. The authors reported genetic results
showing sufficient on-target human DNA (>1%) with typical aDNA damage (>20% C to
T reference mismatches in the first base of the 5' ends of reads) for downstream
population-genetic analysis and Y-chromosome coverage indicative that the three
individuals were genetically male. In addition, one individual (Lab ID: SJN003)
contained DNA suggesting a possible infection by _Treponema pallidum_, a species
with a variety of strains that can cause diseases such as syphilis, bejel and
yaws, and a second individual (Lab ID: SJN001) displayed reads similar to the
Hepatitis B virus. Both results were confirmed by the authors via in-solution
enrichment approaches.

We were able to successfully replicate the human and pathogen screening results
in a single run of nf-core/eager. Mapping to the human reference genome (hs37d5)
with BWA aln and binning of off-target reads with MALT to the NCBI Nucleotide
database (2017-10-26) yielded the same results of all individuals having a
biological sex of male, as well as the same frequency of C to T miscoding
lesions and short fragment lengths (both characteristic of true aDNA).
Metagenomic hits to both pathogens from the corresponding individuals that also
yielded complete genomes in the original publication were also detected. Both
results and other processing statistics were identified via a single interactive
MultiQC report, excerpts of which can be seen in Figure
{@fig:dual-human-pathogen}. The full interactive report can be seen in the
supplementary information.

![Sections of a MultiQC report (v1.10dev) with the outcome of simultaneous human
DNA and microbial pathogen screening with nf-core/eager, including **A**
Sex.DetERRmine output of biological sex assignment with coverages on X and Y
being half of that of autosomes, indicative of male individuals, and **B** HOPS
output with positive detection of both _Treponema pallidum_ and Hepatitis B
virus reads - indicated with blue boxes. Other taxa in HOPS output represent
typical environmental contamination and oral commensal microbiota found in
teeth. Data was shotgun data from Barquera et al. 2020
[@doi:10.1016/j.cub.2020.04.002], and replicated results here were originally
verified in the publication via enrichment methods. The full interactive reports
for both MultiQC v1.9 and v1.10 (see methods) can be seen in the supplementary
information.](images/dual_humanpathogen_exportedmqc_base_edit.png){#fig:dual-human-pathogen
width="80%"}

## Accessibility

Alongside the interactive MultiQC report, we have written extensive
documentation on all parts of running and interpreting the output of the
pipeline. Given that a large fraction of aDNA researchers come from fields
outside computational biology, and thus may have limited computational training,
we have written documentation [@url:https://nf-co.re/eager/docs] that also gives
guidance on how to interpret each section of the report in the context of
high-throughput sequencing data, but with with a special focus on aDNA. This
includes best practice or expected output schematic images that are published
under CC-BY licenses to allow for use in other training material (an example can
be seen in Figure {@fig:doc-image-example}). We hope this open-access resource
will make the study of aDNA more accessible to researchers new to the field, by
providing practical guidelines on how to evaluate characteristics and effects of
aDNA on downstream analyses.

![Example schematic images of pipeline output documentation that can assist new
users in the interpretation of high-throughput sequencing aDNA
processing.](images/fastqc_adapter_content.png){#fig:doc-image-example
width="70%"}

The development of nf-core/eager in Nextflow and the nf-core initiative will
also improve open-source development, while ensuring the high quality of
community contributions to the pipeline. While Nextflow is written primarily in
Groovy, the Nextflow DSL simplifies a number of concepts to an intermediate
level that bioinformaticians without Java/Groovy experience can easily access
(regardless of own programming language experience). Furthermore, Nextflow
places ubiquitous and more widely known command-line interfaces, such as bash,
in a prominent position within the code, rather than custom Java code and
classes (as in EAGER v1). We hope this will motivate further bug fixes and
feature contributions from the community, to keep the pipeline state-of-the-art
and ensure a longer life-cycle. This will also be supported by the open and
active nf-core community who provide general guidance and advice on developing
Nextflow and nf-core pipelines.

## Comparisons with other pipelines

The scope of nf-core/eager is as a generic, initial data processing and
screening tool, and not to act as a tool for performing more experimental
analyses that requires extensive parameter testing such as modelling. As such,
while similar pipelines designed for aDNA have also been released, such as ATLAS
[@doi:10.1101/105346], these generally have been designed with specific contexts
in mind (e.g. human population genetics). We therefore have opted to not include
common downstream analysis such as Principal Component Analysis for population
genetics, or phylogenetic analysis for microbial genomics, but rather focus on
ensuring nf-core/eager produces useful files that can be easily used as input
for common but more experimental and specialised downstream analysis.

Therefore, we compared pipeline run-times of two functionally equivalent and
previously published pipelines to show that the new implementation of
nf-core/eager is equivalent or more efficient than EAGER or PALEOMIX.

| Category        | Functionality                              | EAGER | PALEOMIX | nf-core/eager |
|-----------------|--------------------------------------------|:-----:|:--------:|:-------------:|
| Infrastructure  | Reproducible software environments offered | ✓     | ✗        | ✓             |
|                 | HPC scheduler integration                  | ✗     | ✗        | ✓             |
|                 | Cloud computing integration                | ✗     | ✗        | ✓             |
|                 | Per-process resource optimisation          | ✗     | ~        | ✓             |
|                 | Pipeline-step parallelisation              | ✗     | ✓        | ✓             |
|                 | Command line set up                        | ✗     | ✓        | ✓             |
|                 | GUI set up                                 | ✓     | ✗        | ✓             |
| Preprocessing   | Sequencing lane merging                    | ✓     | ✓        | ✓             |
|                 | Sequencing quality control                 | ✓     | ✗        | ✓             |
|                 | Sequencing artefact removal                | ✗     | ✗        | ✓             |
|                 | Adapter clipping/read merging              | ✓     | ✓        | ✓             |
|                 | Post-processing sequencing QC              | ✗     | ✗        | ✓             |
| Alignment       | Reference mapping                          | ✓     | ✓        | ✓             |
|                 | Reference mapping statistics               | ✓     | ✓        | ✓             |
|                 | Multi-reference mapping                    | ✗     | ✓        | ✗             |
| Postprocessing  | Mapped reads filtering                     | ✓     | ✓        | ✓             |
|                 | Off-target metagenomic profiling           | ✗     | ✗        | ✓             |
|                 | Off-target metagenomic authentication      | ✗     | ✗        | ✓             |
|                 | Library complexity estimation              | ✓     | ✗        | ✓             |
|                 | Duplicate removal                          | ✓     | ✗        | ✓             |
|                 | BAM merging                                | ✗     | ✓        | ✓             |
| Authentication  | Damage read filtering                      | ✓     | ✗        | ✓             |
|                 | Contamination estimation (Human)           | ✓     | ✗        | ✓             |
|                 | Biological sex determination (Human)       | ✗     | ✗        | ✓             |
|                 | Genome coverage estimation                 | ✓     | ✓        | ✓             |
|                 | Damage calculation                         | ✓     | ✓        | ✓             |
|                 | Damage rescaling                           | ✗     | ✓        | ~             |
| Downstream      | SNP Calling/Genotyping                     | ✓     | ~        | ✓             |
|                 | Consensus sequence generation              | ✓     | ~        | ✓             |
|                 | Regions of interest statistics             | ~     | ✓        | ✓             |

Table: Comparison of pipeline functionality of common ancient DNA processing
pipelines. Tick represents full functionality, tilde represents partial functionality, and cross represents not implemented. {#tbl:pipeline-comparison}

We ran each pipeline on a subset of Viking-age genomic data of cod (_Gadus
morhua_) from Star et al. 2017 [@doi:10.1073/pnas.1710186114]. This data was
originally run using PALEOMIX, and was re-run here as described, but with the
latest version of PALEOMIX (v1.2.14), and with equivalent settings for the other
two pipelines as close as possible to the original paper (EAGER with v1.92.33,
and nf-core/EAGER with v2.2.0dev, commit
830c22d448441e5e19508c198f530a7656c9f25d). The respective benchmarking
environment and exact pipeline run settings can be seen in the Methods and
Supplementary Information. Two samples each with three Illumina paired-end
sequencing runs were analysed, with adapter clipping and merging
(AdapterRemoval), mapping (BWA aln), duplicate removal (Picard's MarkDuplicates)
and damage profiling (PALEOMIX: mapDamage2, EAGER and nf-core/EAGER:
DamageProfiler) steps being performed. We ran the commands for each tool
sequentially, but repeated these batches of commands 10 times - to account for
variability in the cloud service's IO connection. Run times were measured using
the GNU time tool (v1.7).

|Pipeline                  |Version  |Environment  |real        |sys        |user          |
|:-------------------------|:--------|:------------|:-----------|:----------|:-------------|
|nf-core-eager (optimised) |2.2.0dev |singularity  |105.6 ± 4.6 |13.6 ± 0.7 |1593 ± 79.7   |
|PALEOMIX (optimised)      |1.2.14   |conda        |130.6 ± 8.7 |12 ± 0.7   |1820.2 ± 36.9 |
|nf-core-eager             |2.2.0dev |singularity  |209.2 ± 4.4 |11 ± 0.9   |1407.7 ± 30.2 |
|EAGER                     |1.92.37  |singularity  |224.2 ± 4.9 |22.9 ± 0.3 |1736.3 ± 70.2 |
|PALEOMIX                  |1.2.14   |conda        |314.6 ± 2.9 |10.7 ± 1   |1506.7 ± 14   |

Table: Comparison of run times in minutes between three ancient DNA pipelines.
PALEOMIX and nf-core/eager have additional runs with 'optimised' parameters with
fairer computational resources matching modern multi-threading strategies.
Values represent mean and standard deviation of run times in minutes, calculated
from the output of the GNU time tool. Real: real time, System: cumulative CPU system-task times,
User: cumulative CPU time of all tasks. {#tbl:benchmarking-runtimes}

A summary of runtimes of the benchmarking tests can be seen in Table
{@tbl:benchmarking-runtimes}. nf-core/eager showed fastest runtimes across all
three time metrics when running on default parameters. This
highlights the improved efficiency of nf-core/eager's asynchronous processing
system and per-process resource customisation (here represented by nf-core/eager
defaults designed for typical HPC set ups).

As a more realistic demonstration of modern computing multi-threading set ups,
we also re-ran PALEOMIX with the flag --max-bwa-threads set to 4 (listed in
Table {@tbl:benchmarking-runtimes} as 'optimised'), which is equivalent to a
single BWA aln process of nf-core/eager. This resulted in a much faster run-time
than that of default nf-core/eager, due to the approach of PALEOMIX of mapping
each lane of a library separately, whereas nf-core/eager will map all lanes of a
single library merged together. Therefore, given that each library was split
across three lanes, increasing the threads of BWA aln to 4 resulted in 12 per
library, whereas nf-core/eager only gave 4 (by default) for a single BWA aln process
of one library. While the PALEOMIX approach is valid, we opted to retain the
per-library mapping as it is often the longest running step of high-throughput
sequenced genome-mapping pipelines, and it prevents flooding of HPC scheduling
systems with many long-running jobs. Secondly, if users regularly use multi-lane
data, due to nf-core/eager's fine-granularity control, they can simply modify
nf-core/eager's BWA aln process resources via config files to account for this. When
we optimised parameters that were used for BWA aln multi-threading and multiple
lanes to the same number of BWA aln threads as the optimised PALEOMIX run,
nf-core/eager again displayed faster runtimes.  All metrics including mapped
reads, percentage on-target, mean depth coverage and mean read lengths across
all pipelines were extremely similar across all pipelines and replicates (see
methods and Table {@tbl:benchmarking-results}).


## Conclusion

nf-core/eager is an efficient, portable, and accessible pipeline for processing
and screening ancient (meta)genomic data. This re-implementation of EAGER into
Nextflow and nf-core will improve reproducibility and scalability of rapidly
increasing aDNA datasets, for both large and small laboratories. Extensive
documentation also enables newcomers to the field to get a practical
understanding on how to interpret aDNA in the context of NGS data processing.
Ultimately, nf-core/eager provides easier access to the latest tools and routine
screening analyses commonly used in the field, and sets up the pipeline for
remaining at the forefront of palaeogenetic analysis.


## Methods

### Installation

nf-core/eager requires only three dependencies: Java (version >= 8), Nextflow and
either a functional Conda installation *or* Docker/Singularity engine
installation. A quick installation guide to follow to get started can be found
in the *Quick start* section of the nf-core/eager repository
[@url:https://github.com/nf-core/eager#quick-start].

### Running

After installation, users can run the pipeline using standard test data by
utilising some of the test profiles we provide (e.g. using Docker):

```bash
nextflow run nf-core/eager -r 2.2.0 -profile test_tsv,docker
```

This will download test data automatically (as recorded in the test_tsv
profile), run the pipeline locally with all software tools containerised in a
Docker image. The pipeline will store the output of that run in the default
'./results' folder of the current directory.

The default pipeline settings assumes paired end FASTQ data, and will run:

- FastQC
- AdapterRemoval2 (merging and adapter clipping)
- post-clipping FastQC (for AdapterRemoval2 performance evaluation)
- BWA mapping (with the 'aln' algorithm)
- samtools flagstat (for mapping statistics)
- endorS.py (for endogenous DNA calculation)
- Picard MarkDuplicates (for PCR amplicon deduplication)
- PreSeq (for library complexity evaluation)
- DamageProfiler and Qualimap2 (for genome coverage statistics)
- MultiQC pipeline run report

If no additional FASTA indices are given, these will also be generated.

The pipeline is highly configurable and most modules can be turned on-and-off
using different flags at the request of the user, to allow a high level of
customisation to each user's needs. For example, to include metagenomic
screening of off-target reads, and sex determination based on on-target mappings
of pre-clipped single-end data:

```bash
nextflow run nf-core/eager -r 2.2.0 \
-profile conda \
--input '/<path>/<to>/*/*R1*.fastq.gz' --single_end \
--fasta '/<path>/<to>/<reference>.fasta.gz' \
--skip_fastqc --skip_adapterremoval \
--run_bam_filtering --bam_discard_unmapped --bam_unmapped_type 'fastq' \
--run_metagenomic_screening \
--metagenomic_tool 'malt' --database '/<path>/<to>/<malt_database>' \
--run_sexdeterrmine
```

#### Profiles

In addition to private locally defined profiles, we utilise a central
configuration repository to enable users from various institutions to use
pipelines on their particular infrastructure more easily
[@url:https://github.com/nf-core/configs]. There are multiple resources listed
in this repository with information on how to add a user's own institutional
configuration profile with help from the nf-core community. These profiles can
be both generic for all nf-core pipelines, but also customised for specific
pipelines.

Users can customise this infrastructure profile by themselves, with the nf-core
community, or with their local system administrator to make sure that the
pipeline runs successfully, and can then rely on the Nextflow and nf-core
framework to ensure compatibility upon further infrastructure changes. For
example, in order to run the nf-core/eager pipeline at the Max Planck Institute
for the Science of Human History (MPI-SHH) in Jena, users only have to run:

```bash
nextflow run nf-core/eager -r 2.2.0 -profile test_tsv,sdag,shh
```

This runs the testing profile of the nf-core/eager pipeline with parameters
specifically adapted a specific HPC system at the MPI-SHH. In some cases, similar
institutional configs for other institutions may already exist (originally
utilised for different nf-core pipelines), so users need not necessarily write
their own.

#### Inputs

The pipeline can be started using (raw) FASTQ files from sequencing or
pre-mapped BAM files. Additionally, the pipeline requires a FASTA reference
genome. If BAM input is provided, an optional conversion to FASTQ is offered,
otherwise BAM files processing will start from the post-mapping stage.

If users have complex set-ups, e.g. multiple sequencing lanes that require
merging of files, the pipeline can be supplied with a tab separated value (TSV)
file to enable such complex data handling. Both FASTQs and BAMs can be provided
in this set up. FASTQs with the same library name and sequencing chemistry but
sequenced across multiple lanes will be concatenated after adapter removal and
prior mapping. Libraries with different sequencing sequencing kits (paired- vs.
single-end) will be merged after mapping. Libraries with the same sample name
and with the same UDG treatment, will be merged after deduplication. If
libraries with the sample name have different UDG treatment, these will be
merged after the aDNA modification stage (i.e. BAM trimming or PMDtools, if
turned on), prior to genotyping, as shown in Figure {@fig:merging-files}.

![Schematic of different processing and merging points based on the nature of
different libraries, as specified by the metadata of a TSV file. Dashed boxes
represent optional library-specific
processes. Colours refer to each merge points, which occur at certain points
along the pipeline depending on the metadata columns defined in the TSV file. ](images/merging_files.png){#fig:merging-files width="70%"}

As Nextflow will automatically download files from URLs, profiles and/or TSV
files can include links to publicly available data (e.g. the European
Bioinformatics Institutes's ENA FTP server). This assists in reproducibility,
because if profiles or TSV files are uploaded with a publication, a researcher
wishing to re-analyse the data in the same way can use the exact settings and
file merging procedures in the original publication, without having to
reconstruct this from prose.

#### Monitoring

Users can either monitor their pipeline execution with the messages Nextflow
prints to the console while running, or utilise companion tools such as
Nextflow's Tower [@url:https://tower.nf] to monitor their analysis pipeline
during runtime.

#### Output

The pipeline produces a multitude of output files in various file formats, with
a more detailed listing available in the user documentation. These include
metrics, statistical analysis data, and standardised output files (BAM, VCF) for
close inspection and further downstream analysis, as well as a MultiQC report.
If an emailing daemon is set up on the server, the latter can be emailed to
users automatically, when starting the pipeline with a dedicated option
(\-\-email you@yourdomain.org).

### Benchmarking

#### Dual Screening of Human and Microbial Pathogen DNA

Full step-by-step instructions on the set up of the human and pathogen screening
demonstration (including input TSV file) can be seen in the supplementary
information. To demonstrate the efficiency and conciseness of nf-core/eager
pipeline in it's dual role for both human and microbial screening of ancient
material, we replicated the results of Barquera et al. 2020
[@doi:10.1016/j.cub.2020.04.002] using using v2.2.0dev (commit: e7471a78a3 and
Nextflow version: 20.04.1).

The following command was used to run the pipeline on the in-house servers at
the MPI-SHH, including a 2 TB memory node for running MALT against the NCBI Nt
(Nucleotide) database, and therefore the centralised custom profile for this
cluster was used.

```bash
nextflow run nf-core/eager -r dev \
-profile microbiome_screening,sdag,shh \
-with-tower \
--input 'barquera2020_pathogenscreening.tsv' \
--fasta 'ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz' \
--bwaalnn 0.01 \
--bwaalnl 32 \
--run_bam_filtering \
--bam_discard_unmapped \
--bam_unmapped_type fastq \
--dedupper 'markduplicates' \
--run_mtnucratio \
--run_nuclear_contamination \
--run_sexdeterrmine \
--sexdeterrmine_bedfile 'https://github.com/nf-core/test-datasets/raw/eager/reference/Human/1240K.pos.list_hs37d5.0based.bed.gz' \
--run_metagenomic_screening \
--metagenomic_tool malt \
--run_maltextract \
--percent_identity 90 \
--malt_top_percent 1 \
--malt_min_support_mode 'reads' \
--metagenomic_min_support_reads 1 \
--malt_max_queries 100 \
--malt_memory_mode load \
--maltextract_taxon_list 'https://raw.githubusercontent.com/rhuebler/HOPS/external/Resources/default_list.txt' \
--maltextract_filter def_anc \
--maltextract_toppercent 0.01 \
--maltextract_destackingoff \
--maltextract_downsamplingoff \
--maltextract_duplicateremovaloff \
--maltextract_matches \
--maltextract_megansummary \
--maltextract_percentidentity 90.0 \
--maltextract_topalignment \
--database 'malt/databases/indexed/index040/full-nt_2017-10/' \
--maltextract_ncbifiles 'resources/'
```

To include the HOPS results from metagenomic screening in the report, we also
re-ran MultiQC with the upcoming version v1.10 (to be integrated into
nf-core/eager on release). We then installed the development version of MultiQC
(commit: 7584e64) as described in the MultiQC documentation
[@url:https://multiqc.info/], and ran the following command in the results
directory of the nf-core/eager run, using the same configuration file.

```bash
multiqc . -c multiqc_config.yaml -n multiqc1_10.html -o multiqc1_10
```

Until MultiQC v1.10 is released, the HOPS heatmap is exported by nf-core/eager in
the corresponding MaltExtract results directory. Reports from both versions (and the standalone HOPS PDF) can be seen in the supplementary information.

### Pipeline Comparison

Full step-by-step instructions on the set up of the pipeline run-time
benchmarking, including environment and tool versions, can be seen in the
supplementary information. EAGER (v1.92.37) and nf-core/eager (v2.2.0dev,
commit: 830c22d; Nextflow v20.04.1) used the provided pre-built singularity
containers for software environments, whereas for PALEOMIX (v1.2.14) we
generated a custom conda environment (see supplementary information for the
`environmental.yaml` file). Run time comparisons were performed on a 32 CPU (AMD
Opteron 23xx) and 256 GB memory Red Hat QEMU Virtual Machine running the Ubuntu
18.04 operating system (Linux Kernel 4.15.0-112). Resource parameters of each
tool were only modified to specify the maximum available on the server and
otherwise left as default.

The following commands were used for each pipeline, with the commands run 10
times, each after cleaning up reference and results directories using a for
loop. Run times of the run commands themselves were measured using GNU Time.

```bash
## EAGER - description of XML files can be seen in supplementary information
singularity exec \
-B ~/benchmarks/output/EAGER:/data ~/.singularity/cache/EAGER-cache/EAGER-GUI_latest.sif \
eagercli \
/data

## PALEOMIX - description of input YAML files can be seen in supplementary
## information
paleomix bam_pipeline run ~/benchmarks/output/paleomix/makefile_paleomix.yaml

## paleomix optimised - description of input YAML files can be seen in
## supplementary information
paleomix bam_pipeline \
run ~/benchmarks/output/paleomix_optimised/makefile_paleomix.yaml \
--bwa-max-threads 4

## nf-core/eager - description of resources configuration file (-c) can be seen
## in supplementary information
nextflow run nf-core/eager -r dev \
--input ~/benchmarks/output/nfcore-eager-optimised/nfcore-eager_tsv.tsv \
-c ~/.nextflow/pub_eager_vikingfish.conf \
-profile pub_eager_vikingfish_optimised,pub_eager_vikingfish,singularity \
--fasta ~/benchmarks/reference/GCF_902167405.1_gadMor3.0_genomic.fasta \
--outdir ~/benchmarks/output/nfcore-eager-optimised/results/ \
-w ~/benchmarks/output/nfcore-eager-optimised/work/ \
--skip_fastqc \
--skip_preseq \
--run_bam_filtering \
--bam_mapping_quality_threshold 25 \
--bam_discard_unmapped \
--bam_unmapped_type 'discard' \
--dedupper 'markduplicates'

##nf-core/eager optimised - description of resources profile(s) with optimised
## bwa threads setting can be seen in supplementary information
nextflow run nf-core/eager -r dev \
--input ~/benchmarks/output/nfcore-eager-optimised/nfcore-eager_tsv.tsv \
-c ~/.nextflow/pub_eager_vikingfish.conf \
-profile pub_eager_vikingfish_optimised,pub_eager_vikingfish,singularity \
--fasta ~/benchmarks/reference/GCF_902167405.1_gadMor3.0_genomic.fasta \
--outdir ~/benchmarks/output/nfcore-eager-optimised/results/ \
-w ~/benchmarks/output/nfcore-eager-optimised/work/ \
--skip_fastqc \
--skip_preseq \
--run_bam_filtering \
--bam_mapping_quality_threshold 25 \
--bam_discard_unmapped \
--bam_unmapped_type 'discard' \
--dedupper 'markduplicates'
```

Mapping results across all pipelines showed very similar values, with low
variation across replicates as can be seen in Table {@tbl:benchmarking-results}.

|sample_name |category              |EAGER              |nf-core/eager     |PALEOMIX          |
|:-----------|:---------------------|:------------------|:-----------------|:-----------------|
|COD076      |processed_reads       |71388991 ± 0       |71388991 ± 0      |72100142 ± 0      |
|COD092      |processed_reads       |69615709 ± 0       |69615709 ± 0      |70249181 ± 0      |
|COD076      |mapped_qf_reads       |16786467.7 ± 106.5 |16786491.1 ± 89.9 |16686607.2 ± 91.3 |
|COD092      |mapped_qf_reads       |16283216.3 ± 71.3  |16283194.7 ± 37.4 |16207986.2 ± 44.4 |
|COD076      |ontarget_qf           |23.5 ± 0           |23.5 ± 0          |23.1 ± 0          |
|COD092      |ontarget_qf           |23.4 ± 0           |23.4 ± 0          |23.1 ± 0          |
|COD076      |dedupped_mapped_reads |12107264.4 ± 87.8  |12107293.7 ± 69.7 |12193415.8 ± 86.7 |
|COD092      |dedupped_mapped_reads |13669323.7 ± 87.6  |13669328 ± 32.4   |13795703.3 ± 47.9 |
|COD076      |mean_depth_coverage   |0.9 ± 0            |0.9 ± 0           |0.9 ± 0           |
|COD092      |mean_depth_coverage   |1 ± 0              |1 ± 0             |1 ± 0             |
|COD076      |mean_read_length      |49.4 ± 0           |49.4 ± 0          |49.4 ± 0          |
|COD092      |mean_read_length      |48.8 ± 0           |48.8 ± 0          |48.7 ± 0          |

Table: Comparison of common results values of key high-throughput short-read data processing and
mapping steps across the three pipelines. 'qf' stands for mapping-quality
filtered reads, calculated from the output of the GNU time tool. All values
represent mean and standard deviation across 10 replicates of each pipeline.
{#tbl:benchmarking-results}

## Data and software availability

All pipeline code is available on GitHub at
[https://github.com/nf-core/eager](https://github.com/nf-core/eager) and
archived with Zenodo under the DOI
[10.5281/zenodo.1465061](https://doi.org/10.5281/zenodo.1465061). The version of
nf-core/eager that this preprint is based on was the 'dev' branch of the GitHub
repository (2.2.0dev), and was released as v2.2.0. Demonstration data for dual
ancient human and pathogen screening from Barquera et al. [@doi:10.1016/j.cub.2020.04.002] is
publicly available on the European Nucleotide Archive (ENA) under project
accession PRJEB37490. The human reference genome (hs37d5) and screening database
(Nucleotide or 'nt', October 2017) was downloaded from National Center for
Biotechnology Information FTP server. Ancient Cod genomic data from Star et al.
[@doi:10.1073/pnas.1710186114] used for benchmarking is publicly available on
the ENA under project accession PRJEB20524. The _Gadus morhua_ reference genome
NCBI accession ID is: GCF_902167405.1.

This paper was collaboratively written with Manubot
[@doi:10.1371/journal.pcbi.1007128], and supplementary information such as
demonstration and benchmarking environments descriptions and walk-through can be
seen on GitHub at
[https://github.com/apeltzer/eager2-paper/](https://github.com/apeltzer/eager2-paper/) and
the `supplement/` directory.


## Competing Interests

No competing interests are declared.


## Acknowledgements

We thank the nf-core community for general support and suggestions during the
writing of the pipeline. We also thank Arielle Munters, Hester van Schalkwyk,
Irina Velsko, Katherine Eaton, Luc Venturini, Marcel Keller, Pierre Lindenbaum,
Pontus Skoglund, Raphael Eisenhofer, Torsten Günter, Kevin Lord, Åshild Vågene
for bug reports and feature suggestions. We are grateful to the members of the
Department of Archaeogenetics at the Max Planck Institute for the Science of
Human History who performed beta testing of the pipeline. We thank the aDNA
twitter community for responding to polls regarding design decisions during
development.

The GWDG kindly provided computational infrastructure for benchmarking. We also
want to thank Selina Carlhoff, Maria Spyrou, Elisabeth Nelson, Alexander Herbig and Wolfgang Haak for providing
comments and suggestions on this manuscript, and acknowledge Christina Warinner,
Stephan Schiffels and the Max Planck Society who provided funds for travel to
nf-core events. This project was also supported by the ERC Starting Grant
project (FoodTransforms) ERC-2015-StG 678901 funded by the European Research
Council awarded to Philipp W. Stockhammer (Ludwig Maximilian University,
Munich).


## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs"></div>
