## Methods

### Installation

nf-core/eager requires only three dependencies: a version of Java, Nextflow and
either a functional Conda installation *or* Docker/Singularity container
installation. A quick installation guide to follow to get started can be found
in the *Quick start* section of the nf-core/eager repository
[@url:https://github.com/nf-core/eager#quick-start].

### Running

After installation, users can run the pipeline using standard test data by
utilising some of the test profiles we provide (e.g. using Docker):

```bash
nextflow run nf-core/eager -r 2.2.0 -profile test_tsv,docker
```

This will download test data automatically (as recorded in the test_tsv
profile), run the pipeline locally with all software tools containerised in a
Docker image. The pipeline will store the output of that run in the default
'./results' folder of the current directory.

The default pipeline settings assumes paired end FASTQ data, and will run:

- FastQC
- AdapterRemoval2 (merging and adapter clipping)
- post-clipping FastQC (for AdapterRemoval2 performance evaluation)
- BWA mapping (with the 'aln' algorithm)
- samtools flagstat (for mapping statistics)
- endorS.py (for endogenous DNA calculation)
- DeDup (for PCR amplicon deduplication)
- PreSeq (for library complexity evaluation)
- DamageProfiler and Qualimap2 (for genome coverage statistics)
- MultiQC pipeline run report

If no additional FASTA indices are given, these will also be generated.

The pipeline is highly configurable and most modules can be turned on-and-off
using different flags at the request of the user, to allow a high level of
customisation to each user's needs. For example, to include metagenomic
screening of off-target reads, and sex determination based on on-target mappings
of pre-clipped single-end data:

```bash
nextflow run nf-core/eager -r 2.2.0 \
-profile conda \
--input '/<path>/<to>/*/*R1*.fastq.gz' --single_end \
--fasta '/<path>/<to>/<reference>.fasta.gz' \
--skip_fastqc --skip_adapterremoval \
--run_bam_filtering --bam_discard_unmapped --bam_unmapped_type 'fastq' \
--run_metagenomic_screening \
--metagenomic_tool 'malt' --database '/<path>/<to>/<malt_database>' \
--run_sexdeterrmine
```

#### Profiles

In addition to private locally defined profiles, we utilise a central
configuration repository to enable users from various institutions to use
pipelines on their particular infrastructure more easily
[@url:https://github.com/nf-core/configs]. There are multiple resources listed
in this repository with information on how to add a user's own institutional
configuration profile with help from the nf-core community. These profiles can
be both generic for all nf-core pipelines, but also customised for specific
pipelines.

Users can customise this infrastructure profile by themselves, with the nf-core
community, or with their local system administrator to make sure that the
pipeline runs successfully, and can then rely on the Nextflow and nf-core
framework to ensure compatibility upon further infrastructure changes. For
example, in order to run the nf-core/eager pipeline at the Max Planck Institute
for the Science of Human History (MPI-SHH) in Jena, users only have to run:

```bash
nextflow run nf-core/eager -r 2.2.0 -profile shh_cdag,test
```

This runs the testing profile of the nf-core/eager pipeline with parameters
specifically adapted to the HPC system at the MPI-SHH. In some cases, similar
institutional configs for other institutions may already exist (originally
utilised for different nf-core pipelines), so users need not necessarily write
their own.

#### Inputs

The pipeline can be started using (raw) FASTQ files from sequencing or
pre-mapped BAM files. Additionally, the pipeline requires a FASTA reference
genome. If BAM input is provided, an optional conversion to FASTQ is offered,
otherwise BAM files processing will start from the post-mapping stage.

If users have complex set-ups, e.g. multiple sequencing lanes that require
merging of files, the pipeline can be supplied with a tab separated value (TSV)
file to enable such complex data handling. Both FASTQs and BAMs can be provided
in this set up. FASTQs with the same library name and sequencing chemistry but
sequenced across multiple lanes will be concatenated after adapter removal and
prior mapping. Libraries with different sequencing sequencing kits (paired- vs.
single-end) will be merged after mapping. Libraries with the sample name and with the same UDG treatment,
will be merged after deduplication. If libraries with the sample name have
different UDG treatment, these will be merged after the aDNA modification stage
(i.e. BAM trimming or PMDtools, if turned on), prior to genotyping, as shown in
Figure {@fig:merging-files}.

![Schematic of different processing and merging points based on the nature of
different libraries, as specified by the metadata of a TSV file. Dashed boxes
represent optional library-specific
processes. Colours refer to each merge points, which occur at certain points
along the pipeline depending on the metadata columns defined in the TSV file. ](images/merging_files.png){#fig:merging-files width="70%"}

As Nextflow will automatically download files from URLs, profiles and/or TSV
files can include links to publicly available data (e.g. the ENA FTP server).
This assists in reproducibility, because if profiles or TSV files are uploaded
with a publication, a researcher wishing to re-analyse the data in the same way
can use the exact settings and merging procedures in the original publication,
without having to reconstruct this from prose.

#### Monitoring

Users can either monitor their pipeline execution with the messages Nextflow
prints to the console while running, or utilise companion tools such as
Nextflow's Tower [@url:https://tower.nf] to monitor their analysis pipeline
during runtime.

#### Output

The pipeline produces a multitude of output files in various file formats, with
a more detailed listing available in the user documentation. These include
metrics, statistical analysis data, and standardised output files (BAM, VCF) for
close inspection and further downstream analysis, as well as a MultiQC report.
If an emailing daemon is set up on the server, the latter can be emailed to
users automatically, when starting the pipeline with a dedicated option
(\-\-email you@yourdomain.org).

## Benchmarking

### Dual Screening of Human and Microbial Pathogen DNA

Full step-by-step instructions on the set up of the human and pathogen screening
demonstration (including input TSV file) can be seen in the supplementary
information. To demonstrate the efficiency and conciseness of nf-core/eager
pipeline in it's dual role for both human and microbial screening of ancient
material, we replicated the results of Barquera et al. 2020
[@doi:10.1016/j.cub.2020.04.002] using using v2.2.0dev (commit: e7471a78a3 and
Nextflow version: 20.04.1).

The following command was used to run the pipeline on the in-house servers at
the MPI-SHH, including a 2 TB memory node for running MALT against the NCBI Nt
(Nucleotide) database, and therefore the centralised custom profile for this
cluster was used.

```bash
nextflow run nf-core/eager -r dev \
-profile microbiome_screening,sdag,shh \
-with-tower \
--input 'barquera2020_pathogenscreening.tsv' \
--fasta 'ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz' \
--bwaalnn 0.01 \
--bwaalnl 32 \
--run_bam_filtering \
--bam_discard_unmapped \
--bam_unmapped_type fastq \
--dedupper 'markduplicates' \
--run_mtnucratio \
--run_nuclear_contamination \
--run_sexdeterrmine \
--sexdeterrmine_bedfile 'https://github.com/nf-core/test-datasets/raw/eager/reference/Human/1240K.pos.list_hs37d5.0based.bed.gz' \
--run_metagenomic_screening \
--metagenomic_tool malt \
--run_maltextract \
--percent_identity 90 \
--malt_top_percent 1 \
--malt_min_support_mode 'reads' \
--metagenomic_min_support_reads 1 \
--malt_max_queries 100 \
--malt_memory_mode load \
--maltextract_taxon_list 'https://raw.githubusercontent.com/rhuebler/HOPS/external/Resources/default_list.txt' \
--maltextract_filter def_anc \
--maltextract_toppercent 0.01 \
--maltextract_destackingoff \
--maltextract_downsamplingoff \
--maltextract_duplicateremovaloff \
--maltextract_matches \
--maltextract_megansummary \
--maltextract_percentidentity 90.0 \
--maltextract_topalignment \
--database 'malt/databases/indexed/index040/full-nt_2017-10/' \
--maltextract_ncbifiles 'resources/'
```

To include the HOPS results from metagenomic screening in the report, we also re-ran
MultiQC with the upcoming version v1.10 (to be integrated into nf-core/eager on
the former's release). We then installed the development version of MultiQC
(commit: 7584e64) as described in the MultiQC documentation
[@url:https://multiqc.info/], and ran the following command
in the results directory of the nf-core/eager run, using the same configuration
file.

```bash
multiqc . -c multiqc_config.yaml -n multiqc1_10.html -o multiqc1_10
```

Until MultiQC v1.10 is released, the HOPS heatmap is exported by nf-core/eager in
the corresponding MaltExtract results directory. Reports from both versions (and the standalone HOPS PDF) can be seen in the supplementary information.

### Pipeline Comparison

Full step-by-step instructions on the set up of the pipeline run-time
benchmarking, including environment and tool versions, can be seen in the
supplementary information. EAGER (v1.92.37) and nf-core/eager (v2.2.0dev,
commit: 830c22d; Nextflow v20.04.1) used the provided pre-built singularity
containers for software environments, whereas for used PALEOMIX (v1.2.14) we
generated a custom conda environment (see supplementary information for the
`environmental.yaml` file). Run time comparisons were performed on a 32 CPU (AMD
Opteron 23xx) and 256 GB memory Red Hat QEMU Virtual Machine running the Ubuntu
18.04 operating system (Kernel 4.15.0-112). Resource parameters of each tool
were only modified to specify the maximum available on the server and otherwise
left as default.

The following commands were used for each pipeline, with the commands run 10
times, each after cleaning up reference and results directories using a for loop.
Run time was measured using GNU Time.

```bash
## EAGER - description of XML files can be seen in supplementary information
singularity exec \
-B ~/benchmarks/output/EAGER:/data ~/.singularity/cache/EAGER-cache/EAGER-GUI_latest.sif \
eagercli \
/data

## PALEOMIX - description of input YAML files can be seen in supplementary
## information
paleomix bam_pipeline run ~/benchmarks/output/paleomix/makefile_paleomix.yaml

## paleomix optimised - description of input YAML files can be seen in
## supplementary information
paleomix bam_pipeline \
run ~/benchmarks/output/paleomix_optimised/makefile_paleomix.yaml \
--bwa-max-threads 4

## nf-core/eager - description of resources configuration file (-c) can be seen
## in supplementary information
nextflow run nf-core/eager -r dev \
--input ~/benchmarks/output/nfcore-eager-optimised/nfcore-eager_tsv.tsv \
-c ~/.nextflow/pub_eager_vikingfish.conf \
-profile pub_eager_vikingfish_optimised,pub_eager_vikingfish,singularity \
--fasta ~/benchmarks/reference/GCF_902167405.1_gadMor3.0_genomic.fasta \
--outdir ~/benchmarks/output/nfcore-eager-optimised/results/ \
-w ~/benchmarks/output/nfcore-eager-optimised/work/ \
--skip_fastqc \
--skip_preseq \
--run_bam_filtering \
--bam_mapping_quality_threshold 25 \
--bam_discard_unmapped \
--bam_unmapped_type 'discard' \
--dedupper 'markduplicates'

##nf-core/eager optimised - description of resources profile(s) with optimised
## bwa threads setting can be seen in supplementary information
nextflow run nf-core/eager -r dev \
--input ~/benchmarks/output/nfcore-eager-optimised/nfcore-eager_tsv.tsv \
-c ~/.nextflow/pub_eager_vikingfish.conf \
-profile pub_eager_vikingfish_optimised,pub_eager_vikingfish,singularity \
--fasta ~/benchmarks/reference/GCF_902167405.1_gadMor3.0_genomic.fasta \
--outdir ~/benchmarks/output/nfcore-eager-optimised/results/ \
-w ~/benchmarks/output/nfcore-eager-optimised/work/ \
--skip_fastqc \
--skip_preseq \
--run_bam_filtering \
--bam_mapping_quality_threshold 25 \
--bam_discard_unmapped \
--bam_unmapped_type 'discard' \
--dedupper 'markduplicates'
```

Mapping results across all pipelines showed very similar values, with low
variation across replicates as can be seen in Table {@tbl:benchmarking-results}.

|sample_name |category              |EAGER              |nf-core/eager     |PALEOMIX          |
|:-----------|:---------------------|:------------------|:-----------------|:-----------------|
|COD076      |processed_reads       |71388991 ± 0       |71388991 ± 0      |72100142 ± 0      |
|COD092      |processed_reads       |69615709 ± 0       |69615709 ± 0      |70249181 ± 0      |
|COD076      |mapped_qf_reads       |16786467.7 ± 106.5 |16786491.1 ± 89.9 |16686607.2 ± 91.3 |
|COD092      |mapped_qf_reads       |16283216.3 ± 71.3  |16283194.7 ± 37.4 |16207986.2 ± 44.4 |
|COD076      |ontarget_qf           |23.5 ± 0           |23.5 ± 0          |23.1 ± 0          |
|COD092      |ontarget_qf           |23.4 ± 0           |23.4 ± 0          |23.1 ± 0          |
|COD076      |dedupped_mapped_reads |12107264.4 ± 87.8  |12107293.7 ± 69.7 |12193415.8 ± 86.7 |
|COD092      |dedupped_mapped_reads |13669323.7 ± 87.6  |13669328 ± 32.4   |13795703.3 ± 47.9 |
|COD076      |mean_depth_coverage   |0.9 ± 0            |0.9 ± 0           |0.9 ± 0           |
|COD092      |mean_depth_coverage   |1 ± 0              |1 ± 0             |1 ± 0             |
|COD076      |mean_read_length      |49.4 ± 0           |49.4 ± 0          |49.4 ± 0          |
|COD092      |mean_read_length      |48.8 ± 0           |48.8 ± 0          |48.7 ± 0          |

Table: Comparison of results values of key NGS data processing and mapping
steps. 'qf' stands for mapping-quality filtered reads, calculated from the
output of the GNU time tool. All values represent mean and standard deviation
across 10 replicates of each pipeline.  {#tbl:benchmarking-results}

## Data and software availability

All pipeline code is available on github at
[https://github.com/nf-core/eager](https://github.com/nf-core/eager) and
archived with Zenodo under the DOI
[10.5281/zenodo.1465061](https://doi.org/10.5281/zenodo.1465061). The version of
nf-core/eager that this preprint is based on was the 'dev' branch of the GitHub
repository (2.2.0dev), and was released as v2.2.0. Demonstration data for dual
ancient human and pathogen screening from [@doi:10.1016/j.cub.2020.04.002] is
publicly available on the European Nucleotide Archive (ENA) under project
accession PRJEB37490. The human reference genome (hs37d5) and screening database
(Nucleotide or 'nt', October 2017) was downloaded from National Center for
Biotechnology Information FTP server. Ancient Cod genomic data from
[@doi:10.1073/pnas.1710186114] used for benchmarking is publicly available on
the ENA under project accession PRJEB20524. The _Gadus morhua_ reference genome
NCBI accession ID is: GCF_902167405.1.

This paper was collaboratively written with Manubot
[@doi:10.1371/journal.pcbi.1007128], and supplementary information such as
demonstration and benchmarking environments descriptions and walk-through can be
seen on Github at
[https://github.com/apeltzer/eager2-paper/](https://github.com/apeltzer/eager2-paper/) and
the `supplement/` directory.
